kernel/uVars.ml

```diff
@@ -107,53 +105,55 @@ struct
     let hashcons (aq, au as a) =
       let qlen = Array.length aq in
       let ulen = Array.length au in
-        if Int.equal qlen 0 && Int.equal ulen 0 then empty
-        else begin
-          for i = 0 to qlen - 1 do
-            let x = Array.unsafe_get aq i in
-            let x' = Quality.hcons x in
-              if x == x' then ()
-              else Array.unsafe_set aq i x'
-          done;
-          for i = 0 to ulen - 1 do
-            let x = Array.unsafe_get au i in
-            let x' = Level.hcons x in
-              if x == x' then ()
-              else Array.unsafe_set au i x'
-          done;
-          a
-        end
+      if Int.equal qlen 0 && Int.equal ulen 0 then 0, empty
+      else begin
+        let h = ref 0 in
+        for i = 0 to qlen - 1 do
+          let x = Array.unsafe_get aq i in
+          let hx, x' = Quality.hcons x in
+          h := Hashset.Combine.combine !h hx;
+          if x == x' then ()
+          else Array.unsafe_set aq i x'
+        done;
+        for i = 0 to ulen - 1 do
+          let x = Array.unsafe_get au i in
+          let hx, x' = Level.hcons x in
+          h := Hashset.Combine.combine !h hx;
+          if x == x' then ()
+          else Array.unsafe_set au i x'
+        done;
+        (* [h] must be positive (XXX why?). *)
+        let h = !h land 0x3FFFFFFF in
+        h, a
+      end
 
     let eq t1 t2 =
       CArray.equal (==) (fst t1) (fst t2)
       && CArray.equal (==) (snd t1) (snd t2)
 
-    let hash (aq,au) =
-      let accu = ref 0 in
-        for i = 0 to Array.length aq - 1 do
-          let l = Array.unsafe_get aq i in
-          let h = Quality.hash l in
-            accu := Hashset.Combine.combine !accu h;
-        done;
-        for i = 0 to Array.length au - 1 do
-          let l = Array.unsafe_get au i in
-          let h = Level.hash l in
-            accu := Hashset.Combine.combine !accu h;
-        done;
-        (* [h] must be positive. *)
-        let h = !accu land 0x3FFFFFFF in
-          h
   end
 
   module HInstance = Hashcons.Make(HInstancestruct)
 
   let hcons = Hashcons.simple_hcons HInstance.generate HInstance.hcons ()
 
-  let hash = HInstancestruct.hash
-
-  let share a = (hcons a, hash a)
-
-  let empty = hcons empty
+  let hash (aq,au) =
+    let accu = ref 0 in
+    for i = 0 to Array.length aq - 1 do
+      let l = Array.unsafe_get aq i in
+      let h = Quality.hash l in
+      accu := Hashset.Combine.combine !accu h;
+    done;
+    for i = 0 to Array.length au - 1 do
+      let l = Array.unsafe_get au i in
+      let h = Level.hash l in
+      accu := Hashset.Combine.combine !accu h;
+    done;
+    (* [h] must be positive (XXX why?). *)
+    let h = !accu land 0x3FFFFFFF in
```

Comment:
This constraint is inherited from the OCaml runtime. If you do not force it to be positive, then it will be different on 32-bit and 64-bit platforms. (For example, while 0x4000000 is perfectly representable on both 32-bit and 64-bit worlds, it it is negative in one and positive in the other.) Whether such a property is actually useful in the Coq kernel, I am not sure.

Reply:
It is probably not useful as many of our hashes don't try to respect it, so it probably isn't respected (although I didn't check)

Reply:
That said, this *may* change the folding order of some maps between architectures and maybe we do care about that, e.g. for some proof search.

Reply:
I don't think we ever use maps indexed by these hash values, do we? If we don't, I would just remove the whole masking.

Reply:
We do, there are a few maps that are internally binary-tree-based hash tables.